File: public/python/019-find-first-and-last-position.py
# LeetCode Problem #34: Find First and Last Position of Element in Sorted Array
# Difficulty: Medium
# Link: https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/

import bisect

# ─────────────────────────────────────────────
# APPROACH 1: Linear Scan | O(n) time | O(1) space
# EXPLAIN: Walk once from both ends to find first and last occurrence.
# WHEN: When the array is small or nearly unsorted and you need a fallback without binary search.

def search_range_linear(nums: list[int], target: int) -> list[int]:
    first = last = -1
    for i, v in enumerate(nums):
        if v == target:
            if first == -1:
                first = i
            last = i
    return [first, last]


# ─────────────────────────────────────────────
# APPROACH 2: Two Binary Searches | O(log n) time | O(1) space
# EXPLAIN: Use bisect_left for the first position and bisect_right - 1 for the last.
# WHEN: Optimal — two independent binary searches, each O(log n); use in every production/interview context.

def search_range_binary(nums: list[int], target: int) -> list[int]:
    left  = bisect.bisect_left(nums, target)
    # If target not present, bisect_left lands on a different value
    if left >= len(nums) or nums[left] != target:
        return [-1, -1]
    right = bisect.bisect_right(nums, target) - 1
    return [left, right]


# ─────────────────────────────────────────────
# APPROACH 3: Manual Binary Search | O(log n) time | O(1) space
# EXPLAIN: Implement both left-boundary and right-boundary binary searches from scratch without the bisect module.
# WHEN: When you want to show explicit binary search implementation and handle edge cases manually.

def search_range_manual(nums: list[int], target: int) -> list[int]:
    def find_left(nums: list[int], target: int) -> int:
        lo, hi, pos = 0, len(nums) - 1, -1
        while lo <= hi:
            mid = (lo + hi) // 2
            if nums[mid] == target:
                pos = mid
                hi  = mid - 1        # keep searching left
            elif nums[mid] < target:
                lo = mid + 1
            else:
                hi = mid - 1
        return pos

    def find_right(nums: list[int], target: int) -> int:
        lo, hi, pos = 0, len(nums) - 1, -1
        while lo <= hi:
            mid = (lo + hi) // 2
            if nums[mid] == target:
                pos = mid
                lo  = mid + 1        # keep searching right
            elif nums[mid] < target:
                lo = mid + 1
            else:
                hi = mid - 1
        return pos

    return [find_left(nums, target), find_right(nums, target)]


# ─────────────────────────────────────────────
# APPROACH 4: Single Binary Search with Flag | O(log n) time | O(1) space
# EXPLAIN: One helper accepts a boolean; when find_first=True bias left, otherwise bias right — DRY version.
# WHEN: When you want a compact, single-function solution that avoids code duplication.

def search_range_flag(nums: list[int], target: int) -> list[int]:
    def binary_search(find_first: bool) -> int:
        lo, hi, pos = 0, len(nums) - 1, -1
        while lo <= hi:
            mid = (lo + hi) // 2
            if nums[mid] == target:
                pos = mid
                if find_first:
                    hi = mid - 1
                else:
                    lo = mid + 1
            elif nums[mid] < target:
                lo = mid + 1
            else:
                hi = mid - 1
        return pos

    return [binary_search(True), binary_search(False)]


# ─────────────────────────────────────────────
# APPROACH 5: Recursive Binary Search | O(log n) time | O(log n) space
# EXPLAIN: Implement both boundary searches recursively; stack depth is O(log n).
# WHEN: When a recursive solution is preferred or required to demonstrate recursion.

def search_range_recursive(nums: list[int], target: int) -> list[int]:
    def find_first(lo: int, hi: int) -> int:
        if lo > hi:
            return -1
        mid = (lo + hi) // 2
        if nums[mid] == target:
            left = find_first(lo, mid - 1)
            return left if left != -1 else mid
        elif nums[mid] < target:
            return find_first(mid + 1, hi)
        else:
            return find_first(lo, mid - 1)

    def find_last(lo: int, hi: int) -> int:
        if lo > hi:
            return -1
        mid = (lo + hi) // 2
        if nums[mid] == target:
            right = find_last(mid + 1, hi)
            return right if right != -1 else mid
        elif nums[mid] < target:
            return find_last(mid + 1, hi)
        else:
            return find_last(lo, mid - 1)

    n = len(nums)
    return [find_first(0, n - 1), find_last(0, n - 1)]


# ─────────────────────────────────────────────
if __name__ == "__main__":
    cases = [
        ([5,7,7,8,8,10], 8, [3, 4]),
        ([5,7,7,8,8,10], 6, [-1,-1]),
        ([],              0, [-1,-1]),
        ([1],             1, [0, 0]),
    ]
    for nums, target, expected in cases:
        assert search_range_linear(nums, target)    == expected
        assert search_range_binary(nums, target)    == expected
        assert search_range_manual(nums, target)    == expected
        assert search_range_flag(nums, target)      == expected
        assert search_range_recursive(nums, target) == expected
    print("All tests passed.")

# Made with Bob

File: public/java/019-find-first-and-last-position.java
/*
 * LeetCode Problem #34: Find First and Last Position of Element in Sorted Array
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
 */
import java.util.*;

class Solution {

    // APPROACH 1: Linear Scan | O(n) time | O(1) space
    // EXPLAIN: Walk the array once; update first on the first occurrence and last on every occurrence.
    // WHEN: Fallback for small or unsorted arrays; not acceptable when O(log n) is explicitly required.
    public int[] searchRange_Linear(int[] nums, int target) {
        int first = -1, last = -1;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == target) {
                if (first == -1) first = i;
                last = i;
            }
        }
        return new int[]{first, last};
    }

    // APPROACH 2: Two Binary Searches | O(log n) time | O(1) space
    // EXPLAIN: Run one binary search biased toward the left bound and another toward the right bound.
    // WHEN: Optimal — two independent O(log n) passes; always use this for sorted arrays.
    public int[] searchRange(int[] nums, int target) {
        return new int[]{findFirst(nums, target), findLast(nums, target)};
    }

    private int findFirst(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1, index = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) {
                index = mid;
                hi = mid - 1; // bias left
            } else if (nums[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return index;
    }

    private int findLast(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1, index = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) {
                index = mid;
                lo = mid + 1; // bias right
            } else if (nums[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return index;
    }

    // APPROACH 3: STL-Style (Arrays.binarySearch analogue) | O(log n) time | O(1) space
    // EXPLAIN: Manually implement lower_bound (first position ≥ target) and upper_bound (first position > target).
    // WHEN: When you want to show awareness of the lower/upper_bound library idiom in Java.
    public int[] searchRange_Bounds(int[] nums, int target) {
        int lo = lowerBound(nums, target);
        if (lo >= nums.length || nums[lo] != target) return new int[]{-1, -1};
        int hi = lowerBound(nums, target + 1) - 1;
        return new int[]{lo, hi};
    }

    // Returns first index where nums[i] >= val
    private int lowerBound(int[] nums, int val) {
        int lo = 0, hi = nums.length;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] < val) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    }

    // APPROACH 4: Single Binary Search with Flag | O(log n) time | O(1) space
    // EXPLAIN: One parameterised helper: when findFirst=true bias left; otherwise bias right.
    // WHEN: DRY approach — one function replaces two; clean when code reuse is valued.
    public int[] searchRange_Flag(int[] nums, int target) {
        return new int[]{
            binarySearch(nums, target, true),
            binarySearch(nums, target, false)
        };
    }

    private int binarySearch(int[] nums, int target, boolean findFirst) {
        int lo = 0, hi = nums.length - 1, res = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] == target) {
                res = mid;
                if (findFirst) hi = mid - 1;
                else           lo = mid + 1;
            } else if (nums[mid] < target) lo = mid + 1;
            else hi = mid - 1;
        }
        return res;
    }

    // APPROACH 5: Recursive Binary Search | O(log n) time | O(log n) space
    // EXPLAIN: Implement both boundary searches recursively — count recursion depth as O(log n) stack space.
    // WHEN: When a recursive solution is required or preferred for clarity; stack depth O(log n) is acceptable.
    public int[] searchRange_Recursive(int[] nums, int target) {
        return new int[]{
            findFirstRec(nums, target, 0, nums.length - 1),
            findLastRec(nums, target, 0, nums.length - 1)
        };
    }

    private int findFirstRec(int[] nums, int target, int lo, int hi) {
        if (lo > hi) return -1;
        int mid = lo + (hi - lo) / 2;
        if (nums[mid] == target) {
            int left = findFirstRec(nums, target, lo, mid - 1);
            return left == -1 ? mid : left;
        }
        return nums[mid] < target ? findFirstRec(nums, target, mid + 1, hi)
                                  : findFirstRec(nums, target, lo, mid - 1);
    }

    private int findLastRec(int[] nums, int target, int lo, int hi) {
        if (lo > hi) return -1;
        int mid = lo + (hi - lo) / 2;
        if (nums[mid] == target) {
            int right = findLastRec(nums, target, mid + 1, hi);
            return right == -1 ? mid : right;
        }
        return nums[mid] < target ? findLastRec(nums, target, mid + 1, hi)
                                  : findLastRec(nums, target, lo, mid - 1);
    }
}

// Made with Bob
